### ISSUE TITLE
Multiple webhook calls from Facebook 

### ISSUE THREAD

**Issue by mikeives** [jgorset/facebook-messenger#187]
I'm using your gem for an app (thanks for the great code). I'm now trying to integrate Facebook Lead Ads but Facebook requires that this information use the same webhook (as it only allows for one webhook).

When I test the webhook I can see it goes to my server however nothing happens. I assume that the Facebook::Messenger::Server that I mounted in routes for the webhook ignores it since it is not one of the stated events in the code.

How do I handle other information (Lead Ads, etc.) from Facebook on the same webhook?

Thanks in advance,

Mike

**Comment by progapandist** [comment]
You could add Sinatra to your project and redefine your webhook so it handles both events as a Sinatra method. I’d give it a try

**Comment by mikeives** [comment]
Thanks propagandist that sounds promising. I'm new to Sinatra and haven't been able to find anything specifically on this type of rerouting ... trying to figure out how to do what you suggest.

Right now my routes file has this line in it: 
`mount Facebook::Messenger::Server, at: "bot"`

In Facebook my app currently has this for the webhook:
`https://appservername.com/bot` 

So if I create a Sinatra method called fbwebhook and change the Facebook webhook fires to be https://appservername.com/fbwebhook, what code would I put in to the Sinatra method to then forward the webhook to the current 'bot' route if it is a Facebook Message and to the new 'leadad' route if it is a Lead Ad?

Thanks in advance ...

**Comment by progapandist** [comment]
So, you're using the bot from Rails? Then I take my words back, probably you should not mix Rails and Sinatra :) Sorry I could not be much help

**Comment by mikeives** [comment]
No problem ... thanks for your thoughts on it.

I'm hoping there's some way the bot can be made to distinguish the lead ad webhook based on the parameters so I can run a different method when it fires.

**Comment by jgorset** [comment]
Sorry for the late reply, @mikeives! I presume you've already got this to work somehow, but otherwise I'd look into how `mount` works in Rails and see if you can put something in front of it that would distinguish the requests for you. Another way to go about it would be to register another app for this, with another webhook.

**Comment by mikeives** [comment]
Hey Johannes ... I actually was not able to get this to work.  Do you have any suggestions for resources on mount? I've never done any work with that. 

With respect to your suggestion to register another app, do you mean another Facebook app. If that's what you meant, I'm wondering if Facebook would allow me to do that.

Mike

**Comment by abury** [comment]
Hi @mikeives 

Was just casually browsing through the tickets and saw this. I've been able to get this work, in the same rails app. You essentially have to rebuild a fair bit of the bot's entry conditions yourself, but it is doable.

Spend some time looking into how the gem hooks into the webhook. As of now, I'm not mounting it at all, I've got the following route declarations:

```ruby
post 'facebook_webhooks', to: 'facebook_webhooks#receive'
get 'facebook_webhooks', to: 'facebook_webhooks#verify'
```

And then in the controller:
```ruby
  def receive
    params['entry'.freeze].each do |entry|
      if entry['changes'.freeze]
        process_leadgen
        head 200 and return
      end
      if entry['messaging'.freeze]
        process_incoming_message(entry)
      end
    end
  end

  def verify
    if valid_verify_token?(params['hub.verify_token'])
      render text: params['hub.challenge'], status: 200
    else
      render text: 'Error; wrong verify token', status: 401
    end
  end
```

There's a bit more code involved in generating the HMAC, checking for a valid signature, etc, but all of this can be found in the gem itself.

@jgorset let me know if you want me to add something to the readme about how to do this. I can see it being a common use case 👍

**Comment by jgorset** [comment]
That would be great, @abury! I'd be happy to merge it.

**Comment by mikeives** [comment]
@abury thanks so much for this. For some reason I didn't see your posting until I saw a notification from @jgorset. I'll definitely try this out today.

Thanks again!

**Comment by mikeives** [comment]
Hey @abury ... are you creating a separate controller etc. or are you inserting this into the gem's code? If it is the second one, where are you inserting it? I couldn't tell based on the routes, etc.

Thanks,

Mike

**Comment by abury** [comment]
@mikeives It's a separate controller, just called facebook_webhooks. You don't need to edit the gem at all to make it work.

**Comment by mikeives** [comment]
Thanks @abury. I created a separate controller and changed the routing to go to it. This worked great and sent the traffic to the controller. To process the messages I created an ActiveJob (since my app is still using Rails 4)as follows:

`class Facebook::Bot::ProcessIncomingMessageJob < ActiveJob::Base
  queue_as :default

  def perform(messaging)
    puts 'in Facebook::Bot::ProcessIncomingMessageJob perform'
    Facebook::Messenger::Bot.receive(JSON.parse(messaging))
  end
end
`

However, the first line of this code is causing the error 'uninitialized constant Facebook::Bot (NameError)' which is then causing the server to crash when it boots up. 

I've commented out the routes and mounted it with this line I was using before:

`mount Facebook::Messenger::Server, at: "bot"`

This runs code located in the app/bot/listen.rb file

Thoughts on what I should do?

Thanks again for this,

Mike
